lib/Test/Stream/Formatter/TAP.pm

```diff
@@ -99,6 +116,193 @@ sub _autoflush {
     select $old_fh;
 }
 
+sub event_tap {
+    my $self = shift;
+    my ($e, $num) = @_;
+
+    my $converter = $CONVERTERS{ref($e)};
+
+    $num = undef if $self->{+NO_NUMBERS};
+
+    # Legacy Support for $e->to_tap
+    unless ($converter) {
+        my $legacy = $e->can('to_tap') or return;
+        return if $legacy == \&Test::Stream::Event::to_tap;
+        warn "'$e' implements 'to_tap'. to_tap methods on events are deprecated.\n";
+        return $e->to_tap($num);
+    }
+
+    return $self->$converter($e, $num);
+}
+
+sub _ok_event {
+    my $self = shift;
+    my ($e, $num) = @_;
+
+    # We use direct hash access for performance. OK events are so common we
+    # need this to be fast.
+    my $name  = $e->{name};
+    my $debug = $e->{debug};
+    my $todo  = $e->{todo};
+    $todo = $debug->{todo} unless defined $todo; # $debug->todo is deprecated
+    my $skip  = $debug->{skip}; # Deprecated
+
+    my $out = "";
+    $out .= "not " unless $e->{pass};
+    $out .= "ok";
+    $out .= " $num" if defined $num;
+    $out .= " - $name" if $name;
+
+    if (defined $skip && defined $todo) {
+        $out .= " # TODO & SKIP";
+        $out .= " $todo" if length $todo;
+    }
+    elsif (defined $todo) {
+        $out .= " # TODO";
+        $out .= " $todo" if length $todo;
+    }
+    elsif (defined $skip) {
+        $out .= " # skip";
+        $out .= " $skip" if length $skip;
+    }
+
+    my @out = [OUT_STD, "$out\n"];
+
+    if ($e->{diag} && @{$e->{diag}}) {
+        my $diag_handle = ($todo || $e->diag_todo || $debug->_no_diag) ? OUT_TODO : OUT_ERR;
+
+        for my $diag (@{$e->{diag}}) {
+            chomp(my $msg = $diag);
+
+            $msg = "# $msg" unless $msg =~ m/^\n/;
+            $msg =~ s/\n/\n# /g;
+            push @out => [$diag_handle, "$msg\n"];
+        }
+    }
+
+    return @out;
+}
+
+sub _skip_event {
+    my $self = shift;
+    my ($e, $num) = @_;
+
+    my $name  = $e->name;
+    my $debug = $e->debug;
+    my $skip  = $e->reason;
+    my $todo  = $e->todo || $debug->_todo;
+
```

Comment:
The $debug can be eliminated here and give you a lazy lookup as 
```perl
|| $e->debug->_todo
```

Seeing you don't use `$debug` more than once.